# Data Design & Representation Midterm - Xuran (Angela) Wang

In [17]:
import pymysql
import warnings
import requests
import json
import codecs
import time
from tqdm import trange
from bs4 import BeautifulSoup
import os
from os import listdir
from os.path import isfile, join
import pandas as pd
from tqdm import notebook
import re
from sqlalchemy import create_engine

# Question 1 SQL Refresher

## 1a

Install and run SQL on your machine. Use some GUI to see / test that it runs.

## 1b

What data types would you chose to store (i) IP addresses (e.g., “192.168.1.1”), (ii) city names (e.g., “West Sacramento”), and (iii) ZIP codes (e.g., “95605”) in? Why?

(i) IP addresses: VARCHAR
IP addresses are not fixed length and are made up of numbers and special characters.
(ii) City names: VARCHAR
City names are made up of a series of characters and are not fixed length. 
(iii) Zip code: CHAR 
Although zip codes are made up of integers, we don’t usually treat zip codes as numbers. We cannot add or substract zip code. Zip codes in United States have a fixed length.

## 1c

Write a small program in Python or Java that (i) connects to your local SQL instance, (ii) creates the database “msba”, (iii) creates the table “ip_addresses” in “msba” containing the columns “ip”, “city”, and “zip”. This small code will create an empty table.

In [2]:
# ignore warnings
warnings.filterwarnings("ignore")
SQL_DB = "msba"
TERM = "blog"


def main():
    # Create SQL table
    SQL_TABLE = "ip_addresses"
    SQL_TABLE_DEF = "(" + \
            "ip VARCHAR(100)" + \
            ",city VARCHAR(100)" + \
            ",zip CHAR(100)" + \
            ")"

    create_sql_table(SQL_TABLE, SQL_TABLE_DEF)

def create_sql_table(SQL_TABLE, SQL_TABLE_DEF):
    try:
        # Connect to server
        conn = pymysql.connect(host='localhost', user = 'root', password = '123456')
        cursor = conn.cursor()

        query = "CREATE DATABASE IF NOT EXISTS " + SQL_DB
        print(query)
        cursor.execute(query);
        
        query = "CREATE TABLE IF NOT EXISTS " + SQL_DB + "." + SQL_TABLE + " " + SQL_TABLE_DEF + ";";
        print(query)
        cursor.execute(query);
        cursor.close()
        conn.close()
        return

    except IOError as e:
        print(e)


if __name__ == "__main__":
    main()

CREATE DATABASE IF NOT EXISTS msba
CREATE TABLE IF NOT EXISTS msba.ip_addresses (ip VARCHAR(100),city VARCHAR(100),zip CHAR(100));


# Question 2 ipstack API

## 2a 

Go to https://ipstack.com/ and make yourself familiar with the API and its use. Request a free API Key from the page.

API Key = aa023f605ac5992552e5bf076391c1ff

## 2b

Specifically, please read “Specify Output Format” and “Specify Response Fields” in https://ipstack.com/documentation , and write down the four URL strings that return the “main” fields in JSON format for IP addresses “8.8.8.8”, “128.120.0.25”, “128.32.12.14”, “64.165.72.144”, and your IP address (use e.g. https://www.vpnmentor.com/tools/ipinfo/).

URL strings:
“8.8.8.8”:
http://api.ipstack.com/8.8.8.8?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main
“128.120.0.25”
http://api.ipstack.com/128.120.0.25?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main
“128.32.12.14”
http://api.ipstack.com/128.32.12.14?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main 
“64.165.72.144” 
http://api.ipstack.com/64.165.72.144?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main
My IP address: 		http://api.ipstack.com/24.4.20.182?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main 


## 2c

Create a program in Python or Java that executes above five API calls and prints the result to screen.: make your code pretty-print the returned JSON. You may use a package to do so.

In [7]:
from pprintpp import pprint as pp

# Execute API calls
response_1 = requests.get("http://api.ipstack.com/8.8.8.8?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main").json()
response_2 = requests.get("http://api.ipstack.com/128.120.0.25?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main").json()
response_3 = requests.get("http://api.ipstack.com/128.32.12.14?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main").json()
response_4 = requests.get("http://api.ipstack.com/64.165.72.144?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main").json()
response_5 = requests.get("http://api.ipstack.com/24.4.20.182?access_key=aa023f605ac5992552e5bf076391c1ff&fields=main").json()

# Pretty-print the returned JSON
for i in range(1,6):
    pp(eval('response_'+str(i)))


{
    'city': 'Mountain View',
    'continent_code': 'NA',
    'continent_name': 'North America',
    'country_code': 'US',
    'country_name': 'United States',
    'ip': '8.8.8.8',
    'latitude': 37.38801956176758,
    'longitude': -122.07431030273438,
    'region_code': 'CA',
    'region_name': 'California',
    'type': 'ipv4',
    'zip': '94041',
}
{
    'city': 'Davis',
    'continent_code': 'NA',
    'continent_name': 'North America',
    'country_code': 'US',
    'country_name': 'United States',
    'ip': '128.120.0.25',
    'latitude': 38.56296157836914,
    'longitude': -121.81600952148438,
    'region_code': 'CA',
    'region_name': 'California',
    'type': 'ipv4',
    'zip': '95616',
}
{
    'city': 'Berkeley',
    'continent_code': 'NA',
    'continent_name': 'North America',
    'country_code': 'US',
    'country_name': 'United States',
    'ip': '128.32.12.14',
    'latitude': 37.87459945678711,
    'longitude': -122.25466918945312,
    'region_code': 'CA',
    'region_n

## 2d

Parse the JSON strings in (c) to an internal Python or Java object for further handling. Then write code that iterates through the five API requests and prints the “city” and “zip” fields to screen.

In [8]:
from types import SimpleNamespace

# Create dictionary
x1 = dict()

# Parse the JSON strings to internal Python object
# 'city', 'zip'
for i in range(1,6):
    x1[i] = json.loads(json.dumps(eval('response_'+str(i))), object_hook=lambda d: SimpleNamespace(**d))
    print('city:', x1[i].city)
    print('zip:', x1[i].zip)
    print()

city: Mountain View
zip: 94041

city: Davis
zip: 95616

city: Berkeley
zip: 94720

city: West Sacramento
zip: 95819

city: South San Francisco
zip: 94080



## 2e

Augment your code from (d) to also write all five “ip”, “city”, and “zip” to the database you created in (1).

In [9]:
# Create dictionary
x2 = dict()

# Parse the JSON strings to internal Python object
# 'ip', 'city', 'zip'
for i in range(1,6):
    x2[i] = json.loads(json.dumps(eval('response_'+str(i))), object_hook=lambda d: SimpleNamespace(**d))
    print('ip:', x2[i].ip)
    print('city:', x2[i].city)
    print('zip:', x2[i].zip)
    print()

ip: 8.8.8.8
city: Mountain View
zip: 94041

ip: 128.120.0.25
city: Davis
zip: 95616

ip: 128.32.12.14
city: Berkeley
zip: 94720

ip: 64.165.72.144
city: West Sacramento
zip: 95819

ip: 24.4.20.182
city: South San Francisco
zip: 94080



In [23]:
# Write all five 'ip', 'city', 'zip' to the database in SQL
def main():
    try:
        # connect to server
        conn = pymysql.connect(host='localhost', user = 'root', password = '123456',database = 'msba')
        cursor = conn.cursor()
        
        query = "INSERT INTO ip_addresses (ip, city, zip) VALUES (%s, %s, %s)"
        value = [("8.8.8.8", "Mountain View", 94041),
               ("128.120.0.25", "Davis", 95616),
               ("128.32.12.14", "Berkeley", 94720),
               ("64.165.72.144", "West Sacramento", 95819),
               ("24.4.20.182", "South San Francisco", 94080)]
        print(query)
        cursor.executemany(query, value)

        conn.commit()
        cursor.close()
        conn.close()  
        print(cursor.rowcount, "records added successfully.")
        
    except IOError as e:
        print(e)
        
if __name__ == "__main__":
    main()

INSERT INTO ip_addresses (ip, city, zip) VALUES (%s, %s, %s)
5 records added successfully.


# Question 3 Web-scraping: President’s Day Discounts on eBay

## 3a

Have a look at eBay’s President’s Day Deals page. https://www.ebay.com/e/daily-deals/hiw-presidents-day-deals-white-sale
What is the GET request's variable name corresponding to the page number? (This section does not involve coding)

_pgn = 

## 3b

Write code that saves all non-sponsored item URLs (the URL you go to when clicking on an item) to the file "deals.txt" in the same directory as your code. (One URL per line)

In [21]:
# Download eBay’s President’s Day Deals page 1-8
for i in trange(1,9):
    headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36",
"Upgrade-Insecure-Requests":"1"} 
    URL = "https://www.ebay.com/e/daily-deals/hiw-presidents-day-deals-white-sale?_pgn=" + str(i)
    page = requests.get(URL, headers=headers)
    name = 'deal'+ str(i) +'.htm'
    time.sleep(10)
    with open(name,'w', encoding='utf-8') as file:
        file.write(page.text)

# Save all non-sponsored item URLs to the file "deals.txt"
list_ = []
for x in range(1,9):
    name = 'deal'+ str(x) +'.htm'
    with open(name, 'r', encoding='utf-8') as file:
        text = file.read()
        soup = BeautifulSoup(text, 'html.parser')
        div = soup.find_all("div", class_= "s-item__info clearfix")
        for i in div:
            links = i.select('a', class_= 's-item__link')
            url = links[0].get('href')
            list_.append(url)
            with open("deals.txt", "w") as output:
                for a in list_:
                    output.write("%s\n" % a)

#len(list_)

100%|████████████████████████████████████████████████████████████████████████████████████| 8/8 [01:27<00:00, 10.92s/it]


## 3c

Write a program that opens the file in (b) and downloads each of the pages (URLs) into the folders "deals". Each file should be named as "<item-id>.html" where you replace "item-id" with the ID of the item you are saving. E.g., "264616053293.html" for the item with ID "264616053293". Note it is always good to put a 4-second pause between queries. Make sure to catch an error and continue if your query runs into problems connecting to eBay (e.g., if your internet is down for 5 seconds, you don't want your entire code to crash).

In [22]:
import os

# Download each of the pages (URLs) into the folders "deals".
with open("deals.txt", "r", encoding='utf-8') as file:
    file = file.read().split("\n")
    for i in trange(0, len(list_)):
        headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36",
"Upgrade-Insecure-Requests":"1"}
        URL = file[i]
        page = requests.get(URL, headers=headers)
        name = str(URL.split('/', 5)[-1].split('?')[0]) + '.htm'
        time.sleep(4)
        dirName = 'C:/Users/Xuran/Desktop/UCD/Data Design & Representation/Midterm/deals'
        with open(os.path.join(dirName, '') + name,'w', encoding='utf-8') as f:
            f.write(page.text)

100%|████████████████████████████████████████████████████████████████████████████████| 184/184 [14:05<00:00,  4.59s/it]


## 3d

Write a separate piece of code that loops through the pages you downloaded in (c) and opens and parses them into a Python or Java xxxxsoup-object. Identify and select:
seller name, seller score, item price, list price, # items sold, title, returns allowed (true / false), shipping price, condition (e.g., used, new, like new, seller refurbished, ...).
In your code, highlight the selector command you choose to obtain each element using comments.

In [82]:
def seller_name(soup):
    seller_name = soup.find_all('span', {'class': 'mbg-nw'})[0].get_text()
    return seller_name  

In [83]:
def seller_score(soup):
    seller_score = re.sub(r'[^\d]', '', soup.find_all('span', {'class': 'mbg-l'})[0].get_text())
    return seller_score

In [84]:
def item_price(soup):
    item_price = None
    for code in (
        # Regular price
        lambda: re.sub(r'[^\d.]', '', soup.find(id='vi-mskumap-none').get_text()),
        # Unit price
        lambda: re.sub(r'[^\d.]', '', soup.find('span', class_='notranslate').get_text())
    ):
        try:
            item_price = code()
        except:
            continue
        break 
        
    else:
        print('Warning: No Item Price Extracted')
        return None
    try:
        item_price = int(float(item_price) * 100)
    except:
        item_price
        
    return item_price 

In [85]:
def list_price(soup):
    try:
        list_price = re.sub(r'[^\d.]', '', soup.find('span', class_= 'notranslate ms-orp').get_text())
    except:
        print('Warning: No List Price Extracted')
        return None
    else:
        list_price1 = int(float(list_price) * 100)
        return list_price1 

In [86]:
def num_items_sold(soup):
    item_sold = soup.find("a", class_="vi-txt-underline")
    try:
        item_sold1 = item_sold.get_text().split(' ')[0]
    except:
        print('Warning: No Number Sold Extracted')
        return None
    else:
        item_sold2 = int(item_sold1.replace(',',''))
        return item_sold2

In [87]:
def title(soup):
    title = soup.find("h1", class_='it-ttl')
    title1 = title.get_text().split('  ')[-1][1:] 
    return title1

In [88]:
def returns_allowed(soup):
    return_allowed = soup.find("span", id= "vi-ret-accrd-txt")
    for i in return_allowed.text:
        check_number = i.isdigit()
        if check_number == True:
            break
    return check_number

In [110]:
def shipping_price(soup):
    shipping = soup.find(id='shSummary').contents[1].get_text().split(' ')[-1]
    
    if 'FREE' in str(shipping).upper():
        return 0
    else:
        return shipping

In [90]:
def item_condition(soup):
    item_condition = soup.find("span", id="vi-cond-addl-info")
    item_condition1 = item_condition.previous_sibling[:-2].split('\t')[-1]
    return item_condition1

In [111]:
from tqdm import notebook

def info(path):
    os.chdir('C:/Users/Xuran/Desktop/UCD/Data Design & Representation/Midterm/deals')
    files = [f for f in listdir(path) if isfile(join(path, f))]
    items = pd.DataFrame(files, columns=['item_id'])
    items['item_id'] = items['item_id'].str.replace('.htm', '')
    
    for file_name in notebook.tqdm(files):
        item_id = re.sub(r'[^\d]', '', file_name)
        with open(file_name, encoding="utf8") as open_file:
            # parse the page and find appropriate tag
            soup = BeautifulSoup(open_file, 'html.parser')
            items.loc[items['item_id'] == item_id, 'seller_name'] = seller_name(soup)
            items.loc[items['item_id'] == item_id, 'seller_score'] = seller_score(soup)
            items.loc[items['item_id'] == item_id, 'item_price'] = item_price(soup)
            items.loc[items['item_id'] == item_id, 'list_price'] = list_price(soup)
            
            if list_price(soup) == None:
                items.loc[items['item_id'] == item_id, 'list_price_available'] = False
            else:
                items.loc[items['item_id'] == item_id, 'list_price_available'] = True
            
            items.loc[items['item_id'] == item_id, 'num_items_sold'] = num_items_sold(soup)
            items.loc[items['item_id'] == item_id, 'title'] = title(soup)
            items.loc[items['item_id'] == item_id, 'returns_allowed'] = returns_allowed(soup)
            items.loc[items['item_id'] == item_id, 'shipping_price'] = shipping_price(soup)    
            items.loc[items['item_id'] == item_id, 'item_condition'] = item_condition(soup)    
            
    return items

In [112]:
path = 'C:/Users/Xuran/Desktop/UCD/Data Design & Representation/Midterm/deals'
items_info = info(path)

## 3e

Use your code snippet from (1) to connect to SQL (either MySQL, MariaDB, or SQLite. Do NOT use SQL GUI or command terminal). Save the information of items in (d) into a single table named "deals" in “msba”. If an item misses ANY of the information in (d), you should insert that missing value as NULL into the table. Convert any price (item price and shipping price) into a "dollar-cent" format (e.g., convert $12.34 into 1234 and $12 into 1200. Make sure the two least significant digits are cents. If an item does not include cents in the price, insert zeros.) and insert the price as INT into the table.

In [113]:
warnings.filterwarnings("ignore")
SQL_DB = "msba"
TERM = "blog"

def main():
    SQL_TABLE = "deals"
    SQL_TABLE_DEF = "(" + \
            "item_id CHAR(12) NOT NULL PRIMARY KEY" + \
            ",seller_name VARCHAR(100)" + \
            ",seller_score VARCHAR(100)" + \
            ",item_price INT" + \
            ",list_price INT" + \
            ",list_price_available BOOLEAN" + \
            ",num_items_sold INT" + \
            ",title VARCHAR(100)" + \
            ",returns_allowed BOOLEAN" + \
            ",shipping_price INT" + \
            ",item_condition VARCHAR(100)" + \
            ")"
    create_sql_table(SQL_TABLE, SQL_TABLE_DEF)

def create_sql_table(SQL_TABLE, SQL_TABLE_DEF):
    try:
        # Connect to server
        conn = pymysql.connect(host='localhost', user = 'root', password = '123456')
        cursor = conn.cursor()

        query = "CREATE DATABASE IF NOT EXISTS " + SQL_DB
        print(query)
        cursor.execute(query);
        
        query = "CREATE TABLE IF NOT EXISTS " + SQL_DB + "." + SQL_TABLE + " " + SQL_TABLE_DEF + ";";
        print(query)
        cursor.execute(query);
        cursor.close()
        conn.close()
        return

    except IOError as e:
        print(e)

if __name__ == "__main__":
    main()
    

CREATE DATABASE IF NOT EXISTS msba
CREATE TABLE IF NOT EXISTS msba.deals (item_id CHAR(12) NOT NULL PRIMARY KEY,seller_name VARCHAR(100),seller_score VARCHAR(100),item_price INT,list_price INT,list_price_available BOOLEAN,num_items_sold INT,title VARCHAR(100),returns_allowed BOOLEAN,shipping_price INT,item_condition VARCHAR(100));


In [114]:
engine = create_engine('mysql://root:Wxr971009!@localhost/msba')
items.to_sql('deals', con=engine, if_exists='replace')

## 3f

In [115]:
print(items_info)

          item_id             seller_name seller_score  item_price  \
0    110841365290               cozyarray       219906       799.0   
1    110924201040               cozyarray       219906       799.0   
2    111634181825               cozyarray       219907      2199.0   
3    111669303786               cozyarray       219906       599.0   
4    111736675898               cozyarray       219906       599.0   
..            ...                     ...          ...         ...   
179  392009676243  www.powersellerusa.com        77178      1499.0   
180  392103349579                howplumb       146985      1497.0   
181  392363657903  www.powersellerusa.com        77178      1299.0   
182  393000070624              dealgenius       712386      1400.0   
183  393017277719              dealgenius       712386      4600.0   

     list_price list_price_available  num_items_sold  \
0        1698.0                 True          2788.0   
1        1798.0                 True         14

In [116]:
items_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 184 entries, 0 to 183
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   item_id               184 non-null    object 
 1   seller_name           184 non-null    object 
 2   seller_score          184 non-null    object 
 3   item_price            184 non-null    float64
 4   list_price            86 non-null     float64
 5   list_price_available  184 non-null    object 
 6   num_items_sold        176 non-null    float64
 7   title                 184 non-null    object 
 8   returns_allowed       184 non-null    object 
 9   shipping_price        184 non-null    float64
 10  item_condition        184 non-null    object 
dtypes: float64(4), object(7)
memory usage: 15.9+ KB


In [117]:
# Seller score
items_info.groupby(['list_price_available','item_condition'])['seller_score'].describe()

count unique     top freq
list_price_available item_condition                          
False                New                8      4   20538    4
                     New with tags     90     24   99658   16
True                 New               11      3   77178    6
                     New with tags     75      9  219906   26

In [118]:
# Item price
items_info.groupby(['list_price_available','item_condition'])['item_price'].describe()

count         mean          std    min  \
list_price_available item_condition                                           
False                New               8.0  1566.000000  1248.832941  550.0   
                     New with tags    90.0  2765.188889  1817.793785  599.0   
True                 New              11.0  2321.272727  2458.091743  999.0   
                     New with tags    75.0  2365.560000  1512.533093  599.0   

                                         25%     50%      75%      max  
list_price_available item_condition                                     
False                New              695.50  1049.0  1923.25   3999.0  
                     New with tags   1498.25  2299.0  3239.00  13499.0  
True                 New             1049.00  1499.0  1998.00   9499.0  
                     New with tags   1399.00  1998.0  2599.00   7981.0

In [119]:
# List price
items_info.groupby(['list_price_available','item_condition'])['list_price'].describe()

count         mean           std     min  \
list_price_available item_condition                                             
False                New               0.0          NaN           NaN     NaN   
                     New with tags     0.0          NaN           NaN     NaN   
True                 New              11.0  7703.363636  11162.132541  2339.0   
                     New with tags    75.0  8007.226667   4136.029770  1698.0   

                                        25%     50%     75%      max  
list_price_available item_condition                                   
False                New                NaN     NaN     NaN      NaN  
                     New with tags      NaN     NaN     NaN      NaN  
True                 New             3032.0  3899.0  6574.0  40595.0  
                     New with tags   4299.0  7998.0  9999.0  24999.0

In [120]:
# Number of items sold
items_info.groupby(['list_price_available','item_condition'])['num_items_sold'].describe()

count         mean           std   min  \
list_price_available item_condition                                           
False                New               8.0  1977.750000   2167.644459  59.0   
                     New with tags    88.0  1410.875000   1584.531504   1.0   
True                 New              10.0  3772.300000   8629.220450  72.0   
                     New with tags    70.0  7716.028571  23838.155313   1.0   

                                        25%     50%      75%       max  
list_price_available item_condition                                     
False                New             127.25  1470.5  2974.00    6273.0  
                     New with tags   247.00   813.5  1883.50    6979.0  
True                 New             172.00   715.0  2580.50   28135.0  
                     New with tags    91.25  1387.0  4784.75  181452.0

In [121]:
# Shipping price
items_info.groupby(['list_price_available','item_condition'])['shipping_price'].describe()

count  mean  std  min  25%  50%  75%  max
list_price_available item_condition                                           
False                New               8.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
                     New with tags    90.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
True                 New              11.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0
                     New with tags    75.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0

## 3g

Use the stats in (f) and tell in your own words, whether the existence of “list price” appear to influence sales. (You do NOT need to run any model or statistic calculations for this part. Make your judgment only by looking at the stats you printed in (f))

The existence of “list price” appears to influence sales. From the summary stats of number of items sold, we can see that sales with list price (mean = 3772 & 7716) is much higher than sales without list price (mean = 1977 & 1410). This make sense because when customers see the comparison between list price and item price, customers will believe: 1) it has a good discount price; 2) the high list (original) price indicates that the item has a good quality. Thus, customers tend to buy the items with list price. 
